usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_main.c

```diff
@@ -0,0 +1,419 @@
+	scsi_hba_open,		/* open */
+	scsi_hba_close,		/* close */
+	nodev,			/* strategy */
+	nodev,			/* print */
+	nodev,			/* dump */
+	nodev,			/* read */
+	nodev,			/* write */
+	smartpqi_ioctl,		/* ioctl */
+	nodev,			/* devmap */
+	nodev,			/* mmap */
+	nodev,			/* segmap */
+	nochpoll,		/* chpoll */
+	ddi_prop_op,		/* cb_prop_op */
+	NULL,			/* streamtab */
+	D_MP,			/* cb_flag */
+	CB_REV,			/* rev */
+	nodev,			/* aread */
+	nodev			/* awrite */
+};
+
+static struct dev_ops smartpqi_ops = {
+	DEVO_REV,		/* dev_rev */
+	0,			/* refcnt */
+	smartpqi_getinfo,	/* info */
+	nulldev,		/* identify */
+	nulldev,		/* probe */
+	smartpqi_attach,	/* attach */
+	smartpqi_detach,	/* detach */
+	nodev,			/* reset */
+	&smartpqi_cb_ops,	/* driver operations */
+	NULL,			/* bus operations */
+	smartpqi_power,		/* power management */
+	smartpqi_quiesce,	/* quiesce */
+};
+
+static struct modldrv modldrv = {
+	&mod_driverops,
+	SMARTPQI_MOD_STRING,
+	&smartpqi_ops,
+};
+
+static struct modlinkage modlinkage = {
+	MODREV_1, &modldrv, NULL
+};
+
+int pqi_do_scan = 0;
+int pqi_do_ctrl = 0;
+int pqi_offline_target = 0;
+int pqi_do_offline = 0;
+
+/*
+ * This is used for data I/O DMA memory allocation. (full 64-bit DMA
+ * physical addresses are supported.)
+ */
+ddi_dma_attr_t smartpqi_dma_attrs = {
+	DMA_ATTR_V0,		/* attribute layout version		*/
+	0x0ull,			/* address low - should be 0 (longlong)	*/
+	0xffffffffffffffffull, /* address high - 64-bit max	*/
+	0x00666600ull,		/* count max - max DMA object size	*/
+	4096,			/* allocation alignment requirements	*/
+	0x78,			/* burstsizes - binary encoded values	*/
+	1,			/* minxfer - gran. of DMA engine	*/
+	0x00666600ull,		/* maxxfer - gran. of DMA engine	*/
+	0x00666600ull,		/* max segment size (DMA boundary)	*/
+	PQI_MAX_SCATTER_GATHER,	/* scatter/gather list length		*/
+	512,			/* granularity - device transfer size	*/
+	0			/* flags, set to 0			*/
+};
+
+ddi_device_acc_attr_t smartpqi_dev_attr = {
+	DDI_DEVICE_ATTR_V1,
+	DDI_STRUCTURE_LE_ACC,
+	DDI_STRICTORDER_ACC,
+	DDI_DEFAULT_ACC
+};
+
+int
+_init(void)
+{
+	int	status;
+
+	if ((status = ddi_soft_state_init(&pqi_state,
+	    sizeof (struct pqi_state), SMARTPQI_INITIAL_SOFT_SPACE)) !=
+	    0) {
+		return (status);
+	}
+
+	if ((status = scsi_hba_init(&modlinkage)) != 0) {
+		ddi_soft_state_fini(&pqi_state);
+		return (status);
+	}
+
+	if ((status = mod_install(&modlinkage)) != 0) {
+		ddi_soft_state_fini(&pqi_state);
+		scsi_hba_fini(&modlinkage);
+	}
+
+	return (status);
+}
+
+int
+_fini(void)
+{
+	int	ret;
+
+	if ((ret = mod_remove(&modlinkage)) == 0) {
+		scsi_hba_fini(&modlinkage);
+		ddi_soft_state_fini(&pqi_state);
+	}
+	return (ret);
+}
+
+/*
+ * The loadable-module _info(9E) entry point
+ */
+int
+_info(struct modinfo *modinfop)
+{
+	/* CONSTCOND */
+	ASSERT(NO_COMPETING_THREADS);
+
+	return (mod_info(&modlinkage, modinfop));
+}
+
+/*ARGSUSED*/
+static int smartpqi_getinfo(dev_info_t *dip, ddi_info_cmd_t cmd, void *arg,
+    void **result)
```

Comment:
I believe that this entry point can be dispensed with for most HBAs -- unless you have your device nodes (not child devices, but for example an HBA specific /devices minor node.  Put succicently, if you don't call ddi_create_minor_node then you probably don't this function.